### DISCUSSION TITLE
[i18n] Allow generation of routes with lowercase locales

### DISCUSSION THREAD

**Discussion by advancingu** [vercel/next.js#31831] [Ideas]
# Feature request

## Is your feature request related to a problem? Please describe.

In Next.js 10.0, internationalized routing generates routes that expose exact locale values into generated routes, e.g. `/en-US`.

## Describe the solution you'd like

For ease of use by end-users and easier legibility, it should be possible to have routes generated with the entire locale part of routes generated in lowercase, e.g. `/en-us`. However, internally the formally correct capitalization e.g. `en-US` should transparently be injected so that retrieving data from external locale-aware systems (e.g. headless CMS) remains easy.

## Describe alternatives you've considered

A (clumsy) alternative is to manually build a folder structure that implements the locale sub-path via dynamic routes, e.g. `/[locale]/[slug].js` where `getStaticPaths()` provides a locale converted to lowercase for `locale` and in `getStaticProps()` the correct capitalization is recalculated from the `locale` prop using custom code.

## Additional context

In cases where the site hosting service considers URLs case-sensitive, having mixed case locales is highly unintuitive for end users and will cause a lot of problems when they manually enter URLs without considering the possibility of case sensitivity.

**Comment by rokinsky** [comment]
Locales are already case insensitive.

**Comment by advancingu** [comment]
Even if the hosting service matches case insensitive, routes are still generated with mixed case which reduces legibility for non-technical users.

**Comment by rokinsky** [comment]
@advancingu you can use any case for the routes you want, e.g:
```js
module.exports = {
  // ...
  i18n: {
    // ...
    locales: [
      "en-us",
      "en-gb",
      "en",
      // ... 
    ],
    // ...
  },
  // ...
};
```

**Comment by HichemBenChaaben** [comment]
@rokinsky 

facing the same issue. 
when adding locale like `en-us` or `en-US` nextjs build fails creating pre-rendered 404 pages

**Comment by rihardsgrislis** [comment]
@rokinsky so, to support both lowercase and uppercase locales we have to set both variants in i18n file?

en-US
en-us

?

Currently we've only en-US specified in i18n configuration and when opening a page with en-us we get a 404 error.

**Comment by joostburg** [comment]
When I visit a lowercase url with a query parameter, NextJs automtically the locale to en-GB example:

`/en-gb/nice-page?q=hello` results in `/en-GB/nice-page?q=hello`

Another strange behaviour with this is when the culture is the default culture, the culture part will be removed from the url.

**Comment by gregghawes** [comment]
Did anything become of this?

**Comment by apexJCL** [comment]
Is there any update?

In my use case, there's a Django server which renders an old version of my app, but Django's translation uses lowercase for the routes (i.e.: `en-gb` instead of `en-GB`). Django cannot be configured to use the proper UTS Locale Identifier format

**Comment by gc7g** [comment]
I'm having the same problem. And can't do redirects because in next.config.js they are case insensitive.
